assume/units/dst_components.py

```diff
@@ -2,782 +2,1593 @@
 
-    @model_part.Constraint(time_steps)
-    def operating_cost_with_el_price(b, t):
+        Returns:
+            pyo.Block: A Pyomo block representing the PV plant with variables and constraints.
         """
-        Calculates the operating cost of the electrolyser based on the electricity price.
 
+        # Define parameters
+        model_block.max_power = pyo.Param(
+            initialize=self.max_power, within=pyo.NonNegativeReals
+        )
+
+        # Define variables
+        model_block.power = pyo.Var(
+            self.time_steps,
+            within=pyo.NonNegativeReals,
+            bounds=(0, self.max_power),
+        )
+        model_block.operating_cost = pyo.Var(self.time_steps, within=pyo.Reals)
+
+        # Define constraints
+
+        # Predefined power profile constraint
+        if self.power_profile is not None:
+            if not isinstance(self.power_profile, pd.Series):
+                raise TypeError(
+                    "Residential PV `power_profile` must be a pandas Series."
+                )
+            if not all(t in self.power_profile.index for t in self.time_steps):
+                raise ValueError(
+                    "All `time_steps` must be present in residential PV `power_profile` index."
+                )
+
+            @model_block.Constraint(self.time_steps)
+            def power_profile_constraint(b, t):
+                """
+                Ensures the PV follows the predefined power profile.
+                """
+                return b.power[t] == self.power_profile[t]
+
+        # Availability profile constraints
+        if self.availability_profile is not None:
+            if not isinstance(self.availability_profile, pd.Series):
+                raise TypeError(
+                    "Residential PV `availability_profile` must be a pandas Series."
+                )
+            if not all(t in self.availability_profile.index for t in self.time_steps):
+                raise ValueError(
+                    "All `time_steps` must be present in residential PV `availability_profile` index."
+                )
+
+            @model_block.Constraint(self.time_steps)
+            def availability_pv_constraint(b, t):
+                """
+                Ensures the PV operates only during available periods.
+                """
+                return b.power[t] <= self.availability_profile[t] * b.max_power
+
+        # Maximum power constraint (redundant due to variable bounds, included for clarity)
+        @model_block.Constraint(self.time_steps)
+        def max_power_pv_constraint(b, t):
+            """
+            Ensures the power output of the PV unit does not exceed the maximum power limit.
+            """
+            return b.power[t] <= b.max_power
+
+        # Operating costs
+        @model_block.Constraint(self.time_steps)
+        def operating_cost_constraint_rule(b, t):
+            return b.operating_cost[t] == b.power[t] * model.electricity_price[t]
```

Comment:
In your comment you mentioned that you did not implement cost constraints for the storage based components (regarding different prices) what if the power out price also differs? Or did you implement it to serve as a base functionality and if it differs it needs to be handled wherever this component gets used?

Reply:
@HafnerMichael yes, this is the diea, that you specify which prices to use in the global energy balance constraint. and the whole system will optimize to that.